cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,529 @@
+    }
+  }
+
+  /**
+   * Check whether this goal is violated after removing the given leader replica. It is violated if there is not be enough
+   * leader replicas on that broker for that topic
+   *
+   * @param replicaToBeRemoved replica to be removed
+   * @return {@code true} if the given replica move would violate the optimized goal (i.e. the move is not acceptable),
+   * {@code false} otherwise.
+   */
+  private boolean doesLeaderReplicaRemoveViolateOptimizedGoal(Replica replicaToBeRemoved) {
+    if (!replicaToBeRemoved.isLeader()) {
+      // Moving a follower replica does not violate/affect this goal
+      return false;
+    }
+    int topicLeaderReplicaCountOnSourceBroker =
+        replicaToBeRemoved.broker().numLeaderReplicasOfTopicInBroker(replicaToBeRemoved.topicPartition().topic());
+    return topicLeaderReplicaCountOnSourceBroker <= minTopicLeadersPerBroker();
+  }
+
+  /**
+   * This is a hard goal; hence, the proposals are not limited to dead broker replicas in case of self-healing.
+   * Sanity Check: The total number of leader replicas of certain topics is sufficient so that these leader replicas
+   * can be moved to satisfy this goal.
+   *
+   * @param clusterModel The state of the cluster.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void initGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    if (_balancingConstraint.topicsWithMinLeadersPerBrokerPattern().pattern().isEmpty()) {
+      _mustHaveLeaderReplicaPerBrokerTopics = Collections.emptySet();
+      return;
+    }
+    _mustHaveLeaderReplicaPerBrokerTopics = Collections.unmodifiableSet(
+        Utils.getTopicNamesMatchedWithPattern(_balancingConstraint.topicsWithMinLeadersPerBrokerPattern(), clusterModel.topics()));
+    // Sanity checks
+    validateTopicsWithMinLeaderReplicaIsNotExcluded(clusterModel, optimizationOptions);
+    validateEnoughLeaderReplicaToDistribute(clusterModel, optimizationOptions);
+    validateBrokersAllowedReplicaMoveExist(clusterModel, optimizationOptions);
+    // Sort leader replicas for each broker based on resource utilization.
+    new SortedReplicasHelper().addPriorityFunc(ReplicaSortFunctionFactory.prioritizeImmigrants())
+                              .trackSortedReplicasFor(replicaSortName(this, false, false), clusterModel);
+  }
+
+  private int minTopicLeadersPerBroker() {
+    return _balancingConstraint.minTopicLeadersPerBroker();
+  }
+
+  private void validateTopicsWithMinLeaderReplicaIsNotExcluded(ClusterModel clusterModel,
+                                                               OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    if (_mustHaveLeaderReplicaPerBrokerTopics.isEmpty() || optimizationOptions.excludedTopics().isEmpty()) {
+      return;
+    }
+    List<String> shouldNotBeExcludedTopics = new ArrayList<>();
```

Comment:
Why do we use a list rather than a set? Does the order matter or do we expect to handle repeated entries?

Reply:
I can change it to use a `set`